modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/validation/OperationReturnTypeModelValidator.java

```diff
@@ -88,9 +92,16 @@ private void validateNonBlockingCallback(MethodElement<? extends Type> operation
     operationMethod.getParameters().stream()
         .filter(p -> p.getType().isSameType(CompletionCallback.class))
         .findFirst().ifPresent(p -> {
-          if (p.getType().getGenerics().isEmpty()) {
+          List<TypeGeneric> generics = p.getType().getGenerics();
+          if (generics.isEmpty()) {
             problemsReporter.addError(new Problem(p, format(MISSING_GENERICS_ERROR_MESSAGE, operationModel.getName(),
                                                             extensionModel.getName(), CompletionCallback.class.getName())));
+          } else {
+            if (generics.get(0).getConcreteType().isSameType(Void.class) &&
```

Comment:
This restriction also applies to blocking operations, and operations that return message collection if it uses Result or List<Result> return type.